source/server/connection_handler_impl.cc

```diff
@@ -84,8 +84,19 @@ void ConnectionHandlerImpl::enableListeners() {
 
 void ConnectionHandlerImpl::ActiveTcpListener::removeConnection(ActiveTcpConnection& connection) {
   ENVOY_CONN_LOG(debug, "adding to cleanup list", *connection.connection_);
-  ActiveTcpConnectionPtr removed = connection.removeFromList(connections_);
+  ActiveConnections& active_connections = connection.active_connections_;
+  ActiveTcpConnectionPtr removed = connection.removeFromList(active_connections.connections_);
   parent_.dispatcher_.deferredDelete(std::move(removed));
+  // Delete map entry only iff connections becomes empty.
+  if (active_connections.connections_.empty()) {
+    auto iter = connections_by_context_.find(&active_connections.filter_chain_);
+    ASSERT(iter != connections_by_context_.end());
+    parent_.dispatcher_.deferredDelete(std::move(iter->second));
```

Comment:
Is this deferred delete explicitly needed? Naively it seems like potentially not? If so can you add a comment on why? Maybe because of the is_deleting_ case? In that case can you just skip all of this logic and let it destroy inline?

Reply:
Yes, right, sorry, we need to refer to the connections wrapper in the connection destructor. Makes sense.